examples/lkj.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+import argparse
+import torch
+
+import pyro
+import pyro.distributions as dist
+import torch.distributions.constraints as constraints
+import math
+from pyro.infer.mcmc import HMC, MCMC, NUTS
+from pyro import poutine
+
+"""
+This simple example is intended to demonstrate how to use an LKJ prior with
+a multivariate distribution.
+
+It generates entirely random, uncorrelated data, and then attempts to fit a correlation matrix
+and vector of variances.
+"""
+
+def model(y):
+    d = y.shape[1]
+    N = y.shape[0]
+    # Vector of variances for each of the d variables
+    theta = pyro.sample("theta", dist.HalfCauchy(torch.full((d,), 1, dtype=y.dtype)))
+    # Lower cholesky factor of a correlation matrix
+    eta = y.new_full((1,), 1)
+    L_omega = pyro.sample("L_omega", dist.CorrLCholeskyLKJPrior(d, eta))
+    # Lower cholesky factor of the covariance matrix
+    L_Omega = torch.mm(torch.diag(theta.sqrt()), L_omega)
+
+    # Vector of expectations
+    mu = y.new_zeros(d)
+
+    with pyro.plate("observations", N) as n:
+        obs = pyro.sample("obs", dist.MultivariateNormal(mu, scale_tril=L_Omega), obs=y[n])
+    return obs
+
+def main(args):
+    y = torch.randn(args.n, args.num_variables).to(dtype=torch.double)
+    if args.cuda:
+        y = y.cuda()
+    nuts_kernel = NUTS(model, jit_compile=args.jit)
```

Comment:
A few things to note - you mentioned about Pyro using really large values when sampling in the unconstrained region, and why that might happen. The way it works is - we start sampling by using an initial sample from the prior, which is transformed to the unconstrained region using a default transform registered for the support (constraint) of the distribution. In this case, the transform is the `CorrLCholeskyTransform`, and the initial sampling starts from what is given by `_inverse`. Subsequent values proposed are from the leapfrog integrator. 

I checked that this example gives a really large initial value for the gradient, i.e. the hamiltonian trajectory diverges for the initial step size (the default I believe is 1) and hence you see this error. If you start from a small step size, you won't face this issue. e.g. in this code, if you change, your invocation of NUTS to `NUTS(model, jit_compile=args.jit, step_size=1e-3)`, the example will run fine without the eps hack. It is possible that there are other numerical stability issues that I haven't looked into yet, but just based on this, I would think that we shouldn't introduce `eps`, since this is probably a corner case that we can either address in HMC, or throw an error asking the user to lower the initial step size, but this doesn't need to a concern of the LKJ prior.

Reply:
Thanks, that explanation helps.  I think the decision about the eps is really a judgment call at the level of the pyro maintainers.   

Let me add two points:

1.  If an out-of-bound value occurs, the result is an unrecoverable LAPACK exception. To me, this suggests that we're better-off enforcing an eps, because of the possibility that somewhere in the sampling some iteration will exceed bounds.  And because otherwise, we will have to advise users that the distribution only works if the step size is reduced as you propose. 

2. If we do choose to keep eps, then the eps has to be larger than it is.  In additional experimentation, I found more cases where, as the matrix grows larger, the eps is insufficient.  My plan (before reading your comment) was to attempt to derive a formula for the minimum eps needed to guarantee that there won't be an LAPACK exception.

Please let me know your preference regarding eps - I'll wait to make the next push until this is resolved.

Reply:
I recommend to verify if your distribution gives right distribution. One way to verify is to sample 100000 samples and test if marginal distribution of each off-diagonal element of correlation matrices (generated by taking `x @ x.t` where x is your sampled Cholesky factors) follows a Beta distribution with a correct mean and variance. I get no numerical issues in my code so I suggest you add more tests to see where is the problem. In case it helps, here are some tests to verify your log_prob implementation is correct:
+ you can test if the log_prob of generated **correlation matrices** are constant when eta=1.
+ When eta != 1, you can test if your implemented log_prob agrees with the log_prob of a transformed distribution, where base distribution is your beta distribution and transform is the transform from partial correlation to cholesky factor (you can use _autograd_log_det function to calculate Jacobian of this transform).

Reply:
@elbamos - Good point. An eps value higher than 1e-4 makes me a bit uncomfortable, and I think it is okay to raise an exception. Later, we can add more context to the error message to allow the user to debug further, e.g. lower step size if they are running HMC; or even add padding if we see a need. I am hesitant to add padding in anticipation of certain errors that might arise, because the padding itself might lead to the kind of issues you had pointed to earlier. The only failure case we know of is when we run NUTS, and I think the right place to address that would be the NUTS code itself (in a later PR). e.g. we can wrap the initial sampling part in a `try catch` block and start with a reasonable step size that doesn't diverge off.

Reply:
@neerajprad Ok, I will take out the eps. 

@fehiepsi You are wrong.  Your code *did* start to throw the same exception, and it *did* fail tests, as soon as the size of the matrix grew to be greater than 5.  

The LKJ prior does *not* follow a Beta distribution.  Individual elements of the z-vector follow Beta distribution*S*, plural.  But that is only true when sampling from the LKJ Prior.  It doesn't give a basis for testing the log probability.

Reply:
I have not touched your distribution implementation so I don't know where is the origin of your issue. The above are just my suggestions.

I just push a branch lkj2 with tests [here](https://github.com/pyro-ppl/pyro/blob/lkj2/tests/distributions/test_lkj.py) for your reference. I have put comments/proofs pretty careful so I think you can understand it. My branch might not work out of box so it just serves as a reference. It does not follow Stan but follow Lewandowski paper. Hope it help!

Reply:
Here is a [gist](https://gist.github.com/fehiepsi/a254d77fcd024c14c674cba02634731c) which I run your example under plate. At you can see, no error at all.